### PR TITLE
Refactor indexer

### PR DESCRIPTION
**Pull request by james-ward**
* Nicer method names
* More helper methods
* Split "firing" state from indexer state machine and add to shooter
* Trivial shooter state machine

### CODE REVIEW

File: robot.py [hunk 1]
```diff
@@ -62,15 +62,16 @@ def createObjects(self):
         self.turret_absolute_encoder = wpilib.DutyCycleEncoder(0)
 
         self.intake_motor = rev.CANSparkMax(9, rev.CANSparkMax.MotorType.kBrushless)
-        # self.intake_piston = wpilib.Solenoid(0, )
-        self.indexer_front_motor = rev.CANSparkMax(
+        self.intake_piston = wpilib.DoubleSolenoid(2, 3)  # TODO check port numbers
```
Comment (OliverW10) [inline robot.py:65]:
```suggestion
        self.intake_piston = wpilib.DoubleSolenoid(wpilib.PneumaticsModuleType.CTREPCM, 2, 3)  # TODO check port numbers
```

File: components/indexer.py [hunk 2]
```diff
@@ -1,56 +1,88 @@
...
+    # The "tunnel" is the horizontal part of the indexer that the cargo enters first
+    # The "chimney" is the vertical section of the indexer that feeds the shooter
+    # The "cat flap" is the moving flap that allows an opposition cargo to be captured
+    indexer_tunnel_motor: rev.CANSparkMax
+    indexer_chimney_motor: rev.CANSparkMax
     colour_sensor: rev.ColorSensorV3
-    prox_sensor1: wpilib.DigitalInput
-    prox_sensor2: wpilib.DigitalInput
-    # indexer_piston: wpilib.Solenoid
+    tunnel_prox_sensor: wpilib.DigitalInput
```
Comment (OliverW10) [inline components/indexer.py:19]: if the second proximity sensor is serving the same purpose as before, would this be better named as lower chimney sensor?

Comment (james-ward) [reply components/indexer.py:19]: Yes, good point.

File: components/indexer.py [hunk 3]
```diff
@@ -1,56 +1,91 @@
...
-        self.indexer_front_motor.setInverted(True)
-        self.indexer_back_motor.setInverted(False)
+        self.indexer_tunnel_motor.setInverted(True)
+        self.indexer_chimney_motor.setInverted(False)
 
     def execute(self) -> None:
-        self.indexer_front_motor.set(self.speeds[0])
-        self.indexer_back_motor.set(self.speeds[1])
+        self.indexer_tunnel_motor.set(self.indexer_speed * self._tunnel_direction)
+        self.indexer_chimney_motor.set(self.indexer_speed * self._chimney_direction)
```
Comment (auscompgeek) [inline components/indexer.py:39]: mypy errors on these lines:
```
components/indexer.py:38: error: Unsupported operand types for * ("float" and "Direction")
components/indexer.py:39: error: Unsupported operand types for * ("float" and "Direction")
```
Have you tried running this in the simulator? :)

```suggestion
        self.indexer_tunnel_motor.set(self.indexer_speed * self._tunnel_direction.value)
        self.indexer_chimney_motor.set(self.indexer_speed * self._chimney_direction.value)
```

File: components/indexer.py [hunk 4]
```diff
@@ -1,56 +1,97 @@
...
-        return (raw.red > raw.blue) == self.is_red
+        is_red_cargo = raw.red > raw.blue
+        return self.has_cargo_in_tunnel() and is_red_cargo == self.are_we_red()
+
+    @feedback
+    def ready_to_intake(self) -> bool:
+        # We cannot have a cargo in the tunnel, and we can't already have two cargo (one in chimney and one trapped)
+        return not self.has_cargo_in_tunnel() and not (
+            self.has_cargo_in_chimney() and self.has_trapped_cargo
+        )
```
Comment (jack-scott) [inline components/indexer.py:80]: Suggesting this as an alternate for readability

```suggestion
        if self.has_cargo_in_tunnel() or  (self.has_cargo_in_chimney() and self.has_trapped_cargo):
             return false

        return true
```

Comment (auscompgeek) [reply components/indexer.py:80]: I'd personally probably go
```suggestion
        return not (
        	self.has_cargo_in_tunnel()
            or (self.has_cargo_in_chimney() and self.has_trapped_cargo)
        )
```

File: components/indexer.py [hunk 5]
```diff
@@ -1,56 +1,137 @@
...
+            self.chimney_speed * self._chimney_direction.value
+        )
+        if self._cat_flap_is_open:
+            self.cat_flap_piston.set(
+                wpilib.DoubleSolenoid.Value.kForward
+            )  # TODO check direction
+        else:
+            self.cat_flap_piston.set(
+                wpilib.DoubleSolenoid.Value.kReverse
+            )  # TODO check direction
```
Comment (LucienMorey) [inline components/indexer.py:63]:
```suggestion
            ) 
```
this was checked

File: components/indexer.py [hunk 6]
```diff
@@ -1,56 +1,137 @@
...
+            self.indexer_tunnel_motor.set(self.tunnel_speed)
+        else:
+            self.indexer_tunnel_motor.set(0.0)
+        self.indexer_chimney_motor.set(
+            self.chimney_speed * self._chimney_direction.value
+        )
+        if self._cat_flap_is_open:
+            self.cat_flap_piston.set(
+                wpilib.DoubleSolenoid.Value.kForward
+            )  # TODO check direction
```
Comment (LucienMorey) [inline components/indexer.py:59]:
```suggestion
            )
```
This was checked

File: components/indexer.py [hunk 7]
```diff
@@ -1,56 +1,137 @@
...
 
-    def set(self, front: int, back: int) -> None:
-        self.speeds = (self.indexer_speed * front, self.indexer_speed * back)
+    def has_red(self) -> bool:
+        return self.colour_sensor.getRawColor().red > 900
 
-    def set_chute(self, on: bool) -> None:
-        # self.indexer_piston.set(on)
-        pass
+    def has_blue(self) -> bool:
```
Comment (auscompgeek) [inline components/indexer.py:136]: These names should refer to where the sensor is so it's obvious what the controller is doing, e.g. `tunnel_has_blue_ball`

**Review by LucienMorey — APPROVED** [review]

File: components/indexer.py [hunk 8]
```diff
@@ -1,56 +1,133 @@
...
+            self.chimney_speed * self._chimney_direction.value
+        )
+        if self._cat_flap_is_open:
+            self.cat_flap_piston.set(wpilib.DoubleSolenoid.Value.kForward)
+        else:
+            self.cat_flap_piston.set(wpilib.DoubleSolenoid.Value.kReverse)
+
+        if self.tunnel_has_red():
+            self.last_colour = Indexer.CargoColour.RED
+        elif self.tunnel_has_blue():
```
Comment (auscompgeek) [inline components/indexer.py:63]: would be prudent to check whether the sensor thinks it's both red and blue, and reject it

**Review by auscompgeek — APPROVED** [review]

**Pull request merged**